include/pisa/linear_quantizer.hpp

```diff
@@ -0,0 +1,20 @@
+#pragma once
+#include <gsl/gsl_assert>
+#include <cmath>
+
+namespace pisa {
+
+struct LinearQuantizer {
+    explicit LinearQuantizer(float max, uint8_t bits) : m_max(max), m_scale(static_cast<float>(1 << (bits)) / max) {
+       Expects(bits <= 32);
```

Comment:
Let's do at least `Expects(bits <= 32 and bits > 0)`. We can also set a min bits if that makes sense. Bit 0 definitely doesn't and can probably do some ugly shit if passed.

Reply:
Now that I think about it, we can easily afford a runtime check here, so we should just throw if incorrect.
```cpp
if (bits > 32 or bits == 0) {
    throw std::runtime_error(fmt::format("Linear quantizer must take a number of bits between 1 and 32 but {} passed", bits));
}
```